### PR TITLE
Casting mode solver matrix to single precision only after it is constructed

### PR DESCRIPTION
**Pull request by momchil-flex**

### CODE REVIEW

File: tidy3d/plugins/mode/solver.py [hunk 1]
```diff
@@ -259,7 +259,7 @@ def conductivity_model_for_pec(eps, threshold=0.9 * pec_val):
         if np.any(eps_offd > TOL_TENSORIAL) or np.any(mu_offd > TOL_TENSORIAL):
```
Comment (lucas-flexcompute) [inline tidy3d/plugins/mode/solver.py:259]: Setting a boolean flag from an if-clause looks like an anti-pattern:
```python
is_tensorial = np.any(eps_offd > TOL_TENSORIAL) or np.any(mu_offd > TOL_TENSORIAL)
```
Also below in `complex_solver`.

File: tidy3d/plugins/mode/solver.py [hunk 2]
```diff
@@ -380,11 +383,13 @@ def any_pec(eps_vec, threshold=0.9 * np.abs(pec_val)):
             print(spl.norm(diff, ord="fro"), spl.norm(aca, ord="fro"), spl.norm(aac, ord="fro"))
 
         # Call the eigensolver. The eigenvalues are -(neff + 1j * keff)**2
+
+        print(mat.dtype, eig_guess.dtype, vec_init.dtype)
```
Comment (lucas-flexcompute) [inline tidy3d/plugins/mode/solver.py:387]: This is probably left from debugging, right?

**Review by lucas-flexcompute — APPROVED** [review]

File: tidy3d/plugins/mode/solver.py (hunk 1 above)
Comment (momchil-flex) [reply tidy3d/plugins/mode/solver.py]: Ah yeah thanks, who knows when this was a leftover from...

File: tidy3d/plugins/mode/solver.py (hunk 2 above)
Comment (momchil-flex) [reply tidy3d/plugins/mode/solver.py]: yep, removed

File: tidy3d/plugins/mode/solver.py [hunk 3]
```diff
@@ -268,41 +286,32 @@ def conductivity_model_for_pec(eps, threshold=0.9 * pec_val):
...
-            )
-
-        return cls.solver_diagonal(eps_tensor, mu_tensor, der_mats, num_modes, vec_init, neff_guess)
+    @classmethod
+    def trim_small_values(cls, mat: sp.csr_matrix, tol: float) -> sp.csr_matrix:
+        """Eliminate elements of matrix ``mat`` for which ``abs(element) / abs(max_element) < tol``.
+        This operates in-place on mat so there is no return.
+        """
+        max_element = np.amax(np.abs(mat))
+        mat.data *= np.abs(mat.data) / max_element > tol
```
Comment (weiliangjin2021) [inline tidy3d/plugins/mode/solver.py:309]: Will it be safer with `mat.data *= np.logical_or(np.abs(mat.data) / max_element > tol, np.abs(mat.data) > tol)`. This doesn't change the number of nonzeros in your example.

Comment (momchil-flex) [reply tidy3d/plugins/mode/solver.py]: Good thinking, I updated this and added a line to changelog, looks good now?

Comment (weiliangjin2021) [reply tidy3d/plugins/mode/solver.py]: All look good to me

**Review by weiliangjin2021 — APPROVED** [review]

**Pull request merged**